src/Pim/Bundle/CatalogBundle/Elasticsearch/Filter/CategoryFilter.php

```diff
@@ -0,0 +1,155 @@
+ */
+class CategoryFilter extends AbstractFieldFilter implements FieldFilterInterface
+{
+    /** @var CategoryRepositoryInterface */
+    protected $categoryRepository;
+
+    /**
+     * @param CategoryRepositoryInterface $categoryRepository
+     * @param array                       $supportedFields
+     * @param array                       $supportedOperators
+     */
+    public function __construct(
+        CategoryRepositoryInterface $categoryRepository,
+        array $supportedFields = [],
+        array $supportedOperators = []
+    ) {
+        $this->supportedFields = $supportedFields;
+        $this->supportedOperators = $supportedOperators;
+        $this->categoryRepository = $categoryRepository;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function addFieldFilter($field, $operator, $value, $locale = null, $scope = null, $options = [])
+    {
+        if ($operator !== Operators::UNCLASSIFIED) {
+            $this->checkValue($field, $value);
+        }
+
+        switch ($operator) {
+            case Operators::IN_LIST:
+                $clause = [
+                    'terms' => [
+                        'categories' => $value
+                    ]
+                ];
+
+                $this->searchQueryBuilder->addFilter($clause);
+                break;
+
+            case Operators::NOT_IN_LIST:
+                $clause = [
+                    'terms' => [
+                        'categories' => $value
+                    ]
+                ];
+
+                $this->searchQueryBuilder->addMustNot($clause);
+                break;
+
+            case Operators::IN_CHILDREN_LIST:
+                $childrenCategoryCodes = $this->getAllChildrenCodes($value);
+                $clause = [
+                    'terms' => [
+                        'categories' => $childrenCategoryCodes
+                    ]
+                ];
+
+                $this->searchQueryBuilder->addFilter($clause);
+                break;
+
+            case Operators::NOT_IN_CHILDREN_LIST:
+                $childrenCategoryCodes = $this->getAllChildrenCodes($value);
+                $clause = [
+                    'terms' => [
+                        'categories' => $childrenCategoryCodes
+                    ]
+                ];
+
+                $this->searchQueryBuilder->addMustNot($clause);
+                break;
+
+            case Operators::UNCLASSIFIED:
+                $clause = [
+                    'exists' => ['field' => 'categories']
+                ];
+                $this->searchQueryBuilder->addMustNot($clause);
+                break;
+
+            case Operators::IN_LIST_OR_UNCLASSIFIED:
+                $shouldClauseInList = [
+                    'terms' => [
+                        'categories' => $value
+                    ],
+                ];
+                $shouldClauseUnclassified = [
+                    'bool' => [
```

Comment:
You need to see the whole query to understand what is going on:
We do a logical OR (with should here), so the final query will look like:
```php
[
  'query' => [
     'bool' => [
       'should' => [
         [
            // here range, terms, exists clause works the fine
            // but should_not operator does not exists and in this particular case I need a "should not exist" or even " should be missing" but missing does not exists.
           // but I cannot simply add a must_not node right below the should.
           // So, I add a "bool" (doing "passe-plat") then must_not
            "bool" => [
                 "must_not" => [
                     "exists" => ["field" => "category"]
                ]
            ]
         ]
    ]
]
```
So, it's not so useless in this precise case.